src/Vault.sol

```diff
@@ -4,38 +4,152 @@ pragma solidity ^0.8.0;
+        IERC20 token = IERC20(asset());
+        require(address(token) != address(0), "Asset token is not set");
+
+        // Transfer tokens directly from vault to market (without burning shares)
+        bool success = token.transfer(msg.sender, amount);
+        require(success, "Token transfer failed");
+
+        emit BorrowedByMarket(msg.sender, amount);
+    }
+
+    // Admin function to repay tokens back to the vault, only callable by the market contract
+    function adminRepay(uint256 amount) external onlyMarket {
+        IERC20 token = IERC20(asset());
+        require(address(token) != address(0), "Asset token is not set");
+        // Transfer tokens from market to vault (without burning shares)
+        bool success = token.transferFrom(msg.sender, address(this), amount);
+        require(success, "Token transfer failed");
+
+        // Emit an event for the repayment action
+        emit RepaidToVault(msg.sender, amount);
+    }
+
+    function totalAssets()
+        public
+        view
+        override
+        onlyValidMarket
+        returns (uint256)
+    {
+        // Retrieves the idle (not lent) assets in the Vault.
+        uint256 idleAssets = totalIdle();
+
+        // Adds the borrowed assets, including interests owed to the platform
+        uint256 borrowedAssets = market._lentAssets();
+
+        // Return the total assets including borrowed amounts
+        return idleAssets + borrowedAssets;
+    }
+
+    function totalIdle() public view returns (uint256) {
+        // Retrieves the idle (not lent) assets in the Vault.
+        return IERC20(asset()).balanceOf(address(this));
+    }
+
+    function maxWithdraw(address user) public view override returns (uint256) {
+        uint256 idleAssets = totalIdle(); // Only available assets in the vault
+        if (idleAssets == 0) return 0; // no assets to withdraw
+
+        uint256 userShares = balanceOf(user); // User's shares
+        uint256 totalShares = totalSupply(); // Total shares issued
+        // User can withdraw a proportion of the idle assets, based on their share ownership
+        return userShares.mulDiv(idleAssets, totalShares, Math.Rounding.Floor);
+    }
+
+    function maxRedeem(address user) public view override returns (uint256) {
+        uint256 sharesBalance = balanceOf(user);
+        uint256 idleAssets = totalIdle();
+
+        if (idleAssets == 0) return 0; // No assets to redeem
+
+        // Convert idle assets to max redeemable shares
+        uint256 maxShares = convertToShares(idleAssets);
+
+        // User can only redeem the lesser of their balance or available shares
+        return sharesBalance < maxShares ? sharesBalance : maxShares;
+    }
+
+    function setMarket(address _market) external onlyOwner {
+        require(address(market) == address(0), "Market already set");
+        require(_market != address(0), "Invalid market address");
+        market = Market(_market);
```

Comment:
we should also check that the market loan asset is the same as this vaults asset. Since you cant change the market afterwards this might lead to an issue otherwise where you pass the wrong market address and brick the vault